dbscan/structref.go

```diff
@@ -18,10 +21,23 @@ func getColumnToFieldIndexMap(structType reflect.Type) map[string][]int {
 	result := make(map[string][]int, structType.NumField())
 	var queue []*toTraverse
 	queue = append(queue, &toTraverse{Type: structType, IndexPrefix: nil, ColumnPrefix: ""})
+
+	visitedTypes := make(map[reflect.Type]int)
 	for len(queue) > 0 {
 		traversal := queue[0]
 		queue = queue[1:]
 		structType := traversal.Type
+
+		// remember visited types and number of visits.
+		if structType.Kind() == reflect.Struct {
+			if visitCount, visited := visitedTypes[structType]; !visited {
+				visitedTypes[structType] = 1
+			} else if visitCount < MaxStructRecursionLevel {
+				visitedTypes[structType] = visitCount + 1
+			} else {
+				continue
+			}
+		}
```

Comment:
The `visitedTypes` variable is global for the whole traversal, it counts all type encounters even is separate branches. We should judge recursion only on types that were visited on the path to the current type, i.e. only the current branch of the traversal. 
I think we should add `visitedTypes` as a field to the `toTraverse` struct, so would every type we iterate over had it's own set of previously visited types.